modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/BulkByScrollUsesAllScrollDocumentsAfterConflictsIntegTests.java

```diff
@@ -0,0 +1,335 @@
+                assertThat(bulkByScrollResponse.getDeleted(), is((long) maxDocs));
+                assertThat(bulkByScrollResponse.getVersionConflicts(), is((long) conflictingUpdates));
+        });
+    }
+
+    <R extends AbstractBulkByScrollRequest<R>,
+     Self extends AbstractBulkByScrollRequestBuilder<R, Self>> void executeConcurrentUpdatesOnSubsetOfDocs(String sourceIndex,
+        String targetIndex,
+        boolean scriptEnabled,
+        AbstractBulkByScrollRequestBuilder<R, Self> requestBuilder,
+        boolean useOptimisticConcurrency,
+        TriConsumer<BulkByScrollResponse, Integer, Integer> resultConsumer) throws Exception {
+        createIndexWithSingleShard(sourceIndex);
+
+        final int numDocs = 100;
+        final int scrollSize = 50;
+        final int maxDocs = 10;
+
+        List<IndexRequestBuilder> indexRequests = new ArrayList<>(numDocs);
+        int noopDocs = 0;
+        for (int i = numDocs; i > 0; i--) {
+            Map<String, Object> source = new HashMap<>();
+            source.put(SORTING_FIELD, i);
+            // Force that the first maxDocs are transformed into a noop
+            if (scriptEnabled && noopDocs < maxDocs) {
+                // Add a marker on the document to signal that this
+                // document should return a noop in the script
+                source.put(RETURN_NOOP_FIELD, true);
+                noopDocs++;
+            }
+            indexRequests.add(client().prepareIndex(sourceIndex).setId(Integer.toString(i)).setSource(source));
+        }
+        indexRandom(true, indexRequests);
+
+        final ThreadPool threadPool = internalCluster().getDataNodeInstance(ThreadPool.class);
+
+        final int writeThreads = threadPool.info(ThreadPool.Names.WRITE).getMax();
+        assertThat(writeThreads, equalTo(1));
+        final EsThreadPoolExecutor writeThreadPool = (EsThreadPoolExecutor) threadPool.executor(ThreadPool.Names.WRITE);
+        final CyclicBarrier barrier = new CyclicBarrier(writeThreads + 1);
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        // Block the write thread pool
+        writeThreadPool.submit(() -> {
+            try {
+                barrier.await();
+                latch.await();
+            } catch (Exception e) {
+                throw new AssertionError(e);
+            }
+        });
+        // Ensure that the write thread blocking task is currently executing
+        barrier.await();
+
+        final SearchResponse searchResponse = client().prepareSearch(sourceIndex)
+            .setSize(scrollSize)
+            .addSort(SORTING_FIELD, SortOrder.DESC)
+            .execute()
+            .actionGet();
+
+        // Modify a subset of the target documents concurrently
+        final List<SearchHit> originalDocs = Arrays.asList(searchResponse.getHits().getHits());
+        final List<SearchHit> docsModifiedConcurrently = new ArrayList<>();
+        if (randomBoolean()) {
```

Comment:
Can we add a case where we run out of docs on the query, i.e., we end up updating less than `max_docs`?

Reply:
Tackled in https://github.com/elastic/elasticsearch/pull/71430/commits/9fe58744570838a106f737225ecb0fbcff6ac17e